Proposals #110 and #122

niyid opened the pull request:
Fully functional and basic Monero (XMR) wallet integrated to Monero RPC
Wallet
running on localhost with the following features:

1. Viewing balance (total and unlocked).
2. Sending XMR.
3. Viewing XMR Primary address to receive XMR.
4. Viewing and searching transactions.
5. Transaction/payment proof/verification.

TODOs:

1. Completion of automation of Monero related trades (integration to
wallet) - Multisig functions (wallets and transactions) will be needed for 2/3 contracts.
2. Unit Tests (first order - XmrFormatter and XmrValidator).
3.Payment proof testing completion.

niyid committed:
Fully functional and basic Monero (XMR) wallet integrated to Monero RPC
Wallet
running on localhost with the following features:

1. Viewing balance (total and unlocked).
2. Sending XMR.
3. Viewing XMR Primary address to receive XMR.
4. Viewing and searching transactions.
5. Transaction/payment proof/verification (almost certified complete).

TODOs:

1. Completion of automation of Monero related trades (integration to
wallet).
2. Unit Tests (first order - XmrFormatter and XmrValidator).
3.Payment proof testing completion.

niyid committed:
Merging with previous revision.

niyid committed:
Completing the switch from Jackson to Gson.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/HttpException.java:
```diff
@@ -0,0 +1,44 @@
+ * 
+ * @author woodser
+ */
+public class HttpException extends RuntimeException {
```
This class can be deleted.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/JsonUtils.java:
```diff
@@ -0,0 +1,193 @@
+						return null;
+					}
+					try {
+						Double doubleVal = in.nextDouble();
```
This can potentially lose precision because it is being downcast to a double.

The [custom type adapter](https://github.com/woodser/monero-java-lite/blob/master/src/main/java/common/utils/GsonBigIntegerTypeAdapter.java) in monero-java-lite resolves this by getting numbers as strings then converting to BigIntegers.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/JsonUtils.java:
```diff
@@ -0,0 +1,193 @@
+ */
+public class JsonUtils {
+
+	// set up jackson object mapper
```
should be updated to "gson"

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/JsonUtils.java:
```diff
@@ -0,0 +1,193 @@
+public class JsonUtils {
+
+	// set up jackson object mapper
+	public static final Gson DEFAULT_MAPPER = new GsonBuilder()
```
The default mapper in JsonUtils should probably not interpret numbers as BigIntegers as it is a general-purpose utility.  The serialize and deserialize methods in JsonUtils accept a custom mapper which probably for the specific use case of interpreting numbers as BigIntegers globally.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/MoneroRpcConnection.java:
```diff
@@ -0,0 +1,236 @@
+  // logger
+  private static final Logger LOGGER = LoggerFactory.getLogger(MoneroRpcConnection.class);
+
+  // custom mapper to deserialize integers to BigIntegers
```
Old comment.  IMO here or JsonUtils.java is a good place to put a custom mapper which maps numbers to BigIntegers.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/MoneroRpcConnection.java:
```diff
@@ -0,0 +1,236 @@
+      // check RPC response for errors
+      validateRpcResponse(respMap, method, params);
+      return respMap;
+    } catch (HttpException e1) {
```
monero-java-lite has been updated to not use HttpException.java which can be deleted.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/MoneroUtils.java:
```diff
@@ -0,0 +1,311 @@
+	 *             if the values cannot be reconciled
+	 */
+	@SuppressWarnings("unchecked")
+	public static <T> T reconcile(T val1, T val2, Boolean resolveDefined, Boolean resolveTrue, Boolean resolveMax) {
```
These methods have been moved to GenUtils.java in monero-java-lite as they are general-purpose utilities.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/MoneroWalletRpc.java:
```diff
@@ -0,0 +1,169 @@
+		return address.getAddress();
+	}
+	
+	public Balance getBalanceData() {
```
These methods and their dependent POJOs duplicate functionality in monero-java-lite.

Long term the XMR integration should use monero-java-lite and these POJOs and implementations should be removed.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/StreamUtils.java:
```diff
@@ -0,0 +1,49 @@
+import java.io.IOException;
+import java.io.InputStream;
+
+import org.apache.commons.io.IOUtils;
```
This dependency needs to be removed.

This class has been updated in monero-java-lite to remove this dependency.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/result/Address.java:
```diff
@@ -0,0 +1,42 @@
+
+import com.google.gson.annotations.Expose;
+
+public class Address implements Serializable {
```
Remove when monero-java-lite project is used.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/result/Balance.java:
```diff
@@ -0,0 +1,82 @@
+import com.google.gson.annotations.Expose;
+import com.google.gson.annotations.SerializedName;
+
+public class Balance implements Serializable {
```
Remove when monero-java-lite project is used.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/result/MoneroTransfer.java:
```diff
@@ -0,0 +1,185 @@
+import com.google.gson.annotations.Expose;
+import com.google.gson.annotations.SerializedName;
+
+public class MoneroTransfer implements Serializable {
```
Remove when monero-java-lite project is used.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/result/MoneroTransferList.java:
```diff
@@ -0,0 +1,76 @@
+
+import com.google.gson.annotations.Expose;
+
+public class MoneroTransferList implements Serializable {
```
Remove when monero-java-lite project is used.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/result/MoneroTx.java:
```diff
@@ -0,0 +1,134 @@
+import com.google.gson.annotations.Expose;
+import com.google.gson.annotations.SerializedName;
+
+public class MoneroTx implements Serializable {
```
Remove when monero-java-lite project is used.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/result/SubAddress.java:
```diff
@@ -0,0 +1,66 @@
+import com.google.gson.annotations.Expose;
+import com.google.gson.annotations.SerializedName;
+
+public class SubAddress implements Serializable {
```
Remove when monero-java-lite project is used.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/result/SubAddressBalance.java:
```diff
@@ -0,0 +1,115 @@
+import com.google.gson.annotations.Expose;
+import com.google.gson.annotations.SerializedName;
+
+public class SubAddressBalance implements Serializable {
```
Remove when monero-java-lite project is used.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/result/SubAddressIndex.java:
```diff
@@ -0,0 +1,41 @@
+
+import com.google.gson.annotations.Expose;
+
+public class SubAddressIndex implements Serializable {
```
Remove when monero-java-lite project is used.

woodser commented on core/src/main/java/bisq/core/xmr/jsonrpc/MoneroWalletRpc.java:
```diff
@@ -0,0 +1,169 @@
+		Map<String, Object> response = rpcConnection.sendJsonRequest("get_address", params);
+		log.debug("response => {}", response);
+		
+		Address address = JsonUtils.DEFAULT_MAPPER.fromJson(JsonUtils.DEFAULT_MAPPER.toJson(response.get("result")), Address.class);
```
The utility methods within JsonUtils.java (`serialize()`, `deserialize()`, `toMap()`, and `fromMap()`) should be used for JSON serialization/deserialization instead of calling GSON_MAPPER `toJson()` or `fromJson()` directly which is implementation-specific and inconsistent with other usage.

niyid replied:
This is resolved.

woodser reviewed (commented):
I have added review comments, summarized here:

- Over 900 lines of code have been added in MoneroWalletRpc.java and its dependent POJOs which are made redundant by the monero-java-lite project and which I do not see formal tests for.  I am questioning if it is worth adding in the first place.  I want to confirm that the plan is to replace this code with code from monero-java-lite?
- This PR uses org.apache.commons.io.IOUtils in StreamUtils.java which cannot be used in Bisq.  StreamUtils.java has been updated to remove this dependency in monero-java-lite.
- JsonUtils.java default mapping should probably not map to BigIntegers.
- The custom mapper which maps to BigIntegers should not lose precision by downcasting to doubles.
- The utility methods within JsonUtils.java (`serialize()`, `deserialize()`, `toMap()`, and `fromMap()`) should be used for JSON serialization/deserialization instead of calling GSON_MAPPER `toJson()` or `fromJson()` directly.

I can merge this PR then we can iterate on these changes (I can help), or these changes can be made before the PR is merged.

Preferences?

niyid commented:
Thanks for the reviews. I will update the PR then you can merge.

Regards.

On Fri, Oct 4, 2019, 18:16 woodser <notifications@github.com> wrote:

> *@woodser* commented on this pull request.
>
> I have added review comments, summarized here:
>
>    - Over 900 lines of code have been added in MoneroWalletRpc.java and
>    its dependent POJOs which are made redundant by the monero-java-lite
>    project and which I do not see formal tests for. I am questioning if it is
>    worth adding in the first place. I want to confirm that the plan is to
>    remove this code when the monero-java-lite project is used?
>    - This PR uses org.apache.commons.io.IOUtils in StreamUtils.java which
>    cannot be used in Bisq. This class has been updated to remove this
>    dependency in StreamUtils.java.
>    - JsonUtils.java default mapping should probably not map to
>    BigIntegers.
>    - The custom mapper which maps to BigIntegers should not lose
>    precision by downcasting to doubles.
>    - The utility methods within JsonUtils.java (serialize(), deserialize(),
>    toMap(), and fromMap()) should be used for JSON
>    serialization/deserialization instead of calling GSON_MAPPER toJson()
>    or fromJson() directly.
>
> I can merge this PR then we can iterate on these changes (I can help), or
> these changes can be made before the PR is merged.
>
> Preferences?
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/HttpException.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331588623>
> :
>
> > @@ -0,0 +1,44 @@
> +package bisq.core.xmr.jsonrpc;
> +
> +/**
> + * Defines an HTTP exception for when HTTP responses are not in the 200s.
> + *
> + * @author woodser
> + */
> +public class HttpException extends RuntimeException {
>
> This class can be deleted.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/JsonUtils.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331589338>
> :
>
> > + *
> + * @author woodser
> + */
> +public class JsonUtils {
> +
> +	// set up jackson object mapper
> +	public static final Gson DEFAULT_MAPPER = new GsonBuilder()
> +			.registerTypeAdapter(BigInteger.class, new TypeAdapter<BigInteger>() {
> +				@Override
> +				public BigInteger read(JsonReader in) throws IOException {
> +					if (in.peek() == JsonToken.NULL) {
> +						in.nextNull();
> +						return null;
> +					}
> +					try {
> +						Double doubleVal = in.nextDouble();
>
> This can potentially lose precision because it is being downcast to a
> double.
>
> The custom type adapter
> <https://github.com/woodser/monero-java-lite/blob/master/src/main/java/common/utils/GsonBigIntegerTypeAdapter.java>
> in monero-java-lite resolves this by getting numbers as strings then
> converting to BigIntegers.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/JsonUtils.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331589443>
> :
>
> > +import com.google.gson.Gson;
> +import com.google.gson.GsonBuilder;
> +import com.google.gson.JsonSyntaxException;
> +import com.google.gson.TypeAdapter;
> +import com.google.gson.stream.JsonReader;
> +import com.google.gson.stream.JsonToken;
> +import com.google.gson.stream.JsonWriter;
> +
> +/**
> + * Collection of utilities for working with JSON.
> + *
> + * @author woodser
> + */
> +public class JsonUtils {
> +
> +	// set up jackson object mapper
>
> should be updated to "gson"
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/JsonUtils.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331590766>
> :
>
> > +import com.google.gson.GsonBuilder;
> +import com.google.gson.JsonSyntaxException;
> +import com.google.gson.TypeAdapter;
> +import com.google.gson.stream.JsonReader;
> +import com.google.gson.stream.JsonToken;
> +import com.google.gson.stream.JsonWriter;
> +
> +/**
> + * Collection of utilities for working with JSON.
> + *
> + * @author woodser
> + */
> +public class JsonUtils {
> +
> +	// set up jackson object mapper
> +	public static final Gson DEFAULT_MAPPER = new GsonBuilder()
>
> The default mapper in JsonUtils should probably not interpret numbers as
> BigIntegers as it is a general-purpose utility. The serialize and
> deserialize methods in JsonUtils accept a custom mapper which probably for
> the specific use case of interpreting numbers as BigIntegers globally.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/MoneroRpcConnection.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331591337>
> :
>
> > +import org.apache.http.entity.StringEntity;
> +import org.apache.http.impl.client.BasicCredentialsProvider;
> +import org.apache.http.impl.client.HttpClients;
> +import org.apache.http.util.EntityUtils;
> +import org.slf4j.Logger;
> +import org.slf4j.LoggerFactory;
> +
> +/**
> + * Maintains a connection and sends requests to a Monero RPC API.
> + */
> +public class MoneroRpcConnection {
> +
> +  // logger
> +  private static final Logger LOGGER = LoggerFactory.getLogger(MoneroRpcConnection.class);
> +
> +  // custom mapper to deserialize integers to BigIntegers
>
> Old comment. IMO here or JsonUtils.java is a good place to put a custom
> mapper which maps numbers to BigIntegers.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/MoneroRpcConnection.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331591619>
> :
>
> > +      // send http request and validate response
> +      HttpPost post = new HttpPost(uri.toString() + "/json_rpc");
> +      HttpEntity entity = new StringEntity(JsonUtils.serialize(body));
> +      post.setEntity(entity);
> +      HttpResponse resp = client.execute(post);
> +      validateHttpResponse(resp);
> +
> +      // deserialize response
> +      Map<String, Object> respMap = JsonUtils.toMap(JsonUtils.DEFAULT_MAPPER, StreamUtils.streamToString(resp.getEntity().getContent()));
> +      LOGGER.debug("Received response to method '" + method + "': " + JsonUtils.serialize(respMap));
> +      EntityUtils.consume(resp.getEntity());
> +
> +      // check RPC response for errors
> +      validateRpcResponse(respMap, method, params);
> +      return respMap;
> +    } catch (HttpException e1) {
>
> monero-java-lite has been updated to not use HttpException.java which can
> be deleted.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/MoneroUtils.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331591943>
> :
>
> > +	 * @param val2
> +	 *            is a value to reconcile
> +	 * @param resolveDefined
> +	 *            uses defined value if true or null, null if false
> +	 * @param resolveTrue
> +	 *            uses true over false if true, false over true if false, must be
> +	 *            equal if null
> +	 * @param resolveMax
> +	 *            uses max over min if true, min over max if false, must be equal if
> +	 *            null
> +	 * @returns the reconciled value if reconcilable
> +	 * @throws Exception
> +	 *             if the values cannot be reconciled
> +	 */
> +	@SuppressWarnings("unchecked")
> +	public static <T> T reconcile(T val1, T val2, Boolean resolveDefined, Boolean resolveTrue, Boolean resolveMax) {
>
> These methods have been moved to GenUtils.java in monero-java-lite as they
> are general-purpose utilities.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/MoneroWalletRpc.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331592609>
> :
>
> > +		this.rpcConnection = rpcConnection;
> +	}
> +	
> +	public String getPrimaryAddress() {
> +		Map<String, Object> params = new HashMap<>();
> +		params.put("account_index", 0);
> +		Map<String, Object> response = rpcConnection.sendJsonRequest("get_address", params);
> +		log.debug("response => {}", response);
> +		
> +		Address address = JsonUtils.DEFAULT_MAPPER.fromJson(JsonUtils.DEFAULT_MAPPER.toJson(response.get("result")), Address.class);
> +		log.debug("address => {}", address);
> +		
> +		return address.getAddress();
> +	}
> +	
> +	public Balance getBalanceData() {
>
> These methods and their dependent POJOs duplicate functionality in
> monero-java-lite.
>
> Long term the XMR integration should use monero-java-lite and these POJOs
> and implementations should be removed.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/StreamUtils.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331593036>
> :
>
> > @@ -0,0 +1,49 @@
> +package bisq.core.xmr.jsonrpc;
> +
> +import java.io.ByteArrayInputStream;
> +import java.io.IOException;
> +import java.io.InputStream;
> +
> +import org.apache.commons.io.IOUtils;
>
> This dependency needs to be removed.
>
> This class has been updated in monero-java-lite to remove this dependency.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/result/Address.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331593594>
> :
>
> > @@ -0,0 +1,42 @@
> +package bisq.core.xmr.jsonrpc.result;
> +
> +import java.io.Serializable;
> +import java.util.List;
> +
> +import com.google.gson.annotations.Expose;
> +
> +public class Address implements Serializable {
>
> Remove when monero-java-lite project is used.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/result/Balance.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331593641>
> :
>
> > @@ -0,0 +1,82 @@
> +package bisq.core.xmr.jsonrpc.result;
> +
> +import java.io.Serializable;
> +import java.math.BigInteger;
> +import java.util.List;
> +
> +import com.google.gson.annotations.Expose;
> +import com.google.gson.annotations.SerializedName;
> +
> +public class Balance implements Serializable {
>
> Remove when monero-java-lite project is used.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/result/MoneroTransfer.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331593710>
> :
>
> > @@ -0,0 +1,185 @@
> +package bisq.core.xmr.jsonrpc.result;
> +
> +import java.io.Serializable;
> +import java.math.BigInteger;
> +
> +import com.google.gson.annotations.Expose;
> +import com.google.gson.annotations.SerializedName;
> +
> +public class MoneroTransfer implements Serializable {
>
> Remove when monero-java-lite project is used.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/result/MoneroTransferList.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331594096>
> :
>
> > @@ -0,0 +1,76 @@
> +package bisq.core.xmr.jsonrpc.result;
> +
> +import java.io.Serializable;
> +import java.util.Collections;
> +import java.util.List;
> +
> +import com.google.gson.annotations.Expose;
> +
> +public class MoneroTransferList implements Serializable {
>
> Remove when monero-java-lite project is used.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/result/MoneroTx.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331594124>
> :
>
> > @@ -0,0 +1,134 @@
> +package bisq.core.xmr.jsonrpc.result;
> +
> +import java.io.Serializable;
> +import java.math.BigInteger;
> +
> +import com.google.gson.annotations.Expose;
> +import com.google.gson.annotations.SerializedName;
> +
> +public class MoneroTx implements Serializable {
>
> Remove when monero-java-lite project is used.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/result/SubAddress.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331594151>
> :
>
> > @@ -0,0 +1,66 @@
> +package bisq.core.xmr.jsonrpc.result;
> +
> +import java.io.Serializable;
> +
> +import com.google.gson.annotations.Expose;
> +import com.google.gson.annotations.SerializedName;
> +
> +public class SubAddress implements Serializable {
>
> Remove when monero-java-lite project is used.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/result/SubAddressBalance.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331594194>
> :
>
> > @@ -0,0 +1,115 @@
> +package bisq.core.xmr.jsonrpc.result;
> +
> +import java.io.Serializable;
> +import java.math.BigInteger;
> +
> +import com.google.gson.annotations.Expose;
> +import com.google.gson.annotations.SerializedName;
> +
> +public class SubAddressBalance implements Serializable {
>
> Remove when monero-java-lite project is used.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/result/SubAddressIndex.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331594251>
> :
>
> > @@ -0,0 +1,41 @@
> +package bisq.core.xmr.jsonrpc.result;
> +
> +import java.io.Serializable;
> +
> +import com.google.gson.annotations.Expose;
> +
> +public class SubAddressIndex implements Serializable {
>
> Remove when monero-java-lite project is used.
> ------------------------------
>
> In core/src/main/java/bisq/core/xmr/jsonrpc/MoneroWalletRpc.java
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1#discussion_r331602472>
> :
>
> > +	
> +	protected Logger log = LoggerFactory.getLogger(MoneroWalletRpc.class);
> +	
> +	private MoneroRpcConnection rpcConnection;
> +
> +	public MoneroWalletRpc(MoneroRpcConnection rpcConnection) {
> +		this.rpcConnection = rpcConnection;
> +	}
> +	
> +	public String getPrimaryAddress() {
> +		Map<String, Object> params = new HashMap<>();
> +		params.put("account_index", 0);
> +		Map<String, Object> response = rpcConnection.sendJsonRequest("get_address", params);
> +		log.debug("response => {}", response);
> +		
> +		Address address = JsonUtils.DEFAULT_MAPPER.fromJson(JsonUtils.DEFAULT_MAPPER.toJson(response.get("result")), Address.class);
>
> The utility methods within JsonUtils.java (serialize(), deserialize(),
> toMap(), and fromMap()) should be used for JSON
> serialization/deserialization instead of calling GSON_MAPPER toJson() or
> fromJson() directly which is implementation-specific and inconsistent
> with other usage.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bisq-network/incubator-bisq-xmr-integration/pull/1?email_source=notifications&email_token=AE2MWR7O4HQLTO5CR2RLGEDQM53AFA5CNFSM4I4OQPU2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCG6NVSQ#pullrequestreview-297589450>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2MWR2OZTCJJOK7GCJVAHLQM53AFANCNFSM4I4OQPUQ>
> .
>

niyid committed:
Code review updates.

niyid committed:
GsonBigIntegerTypeAdapter - Fixed adapter failing for numbers defaulting
to scientific (exponential) notation.
JsonUtils - Methods with no adapter call equivalent methods by passing
the default adapter.

niyid commented:
The README and the docs should contain some information on how to run the **`monero-wallet-rpc`**  correctly for successful connection to Bisq. Is the developer responsible for this or is this handled by a documentation writer?

niyid commented:
> I can merge this PR then we can iterate on these changes (I can help), or these changes can be made before the PR is merged.
> 
> Preferences?
@woodser 

If it is OK with you, please go ahead with the merging. The next technical TODOs in order are:

1. Switch to **`monero-java-lite`**.
2. Implementation of 2/3 multisig.
3. Trade automation.

Regards.

woodser commented:
@niyid Can you please remove the dependency on org.apache.commons.io.IOUtils from StreamUtils.java by updating to the version in monero-java-lite?  That should be an easy change and the last dependency to remove, then I am happy to merge this PR.

niyid committed:
Removed Commons IO dependency.

niyid commented:
> @niyid Can you please remove the dependency on org.apache.commons.io.IOUtils from StreamUtils.java by updating to the version in monero-java-lite? That should be an easy change and the last dependency to remove, then I am happy to merge this PR.

@woodser 

I have removed the dependency.
